client/src/main/java/io/pravega/client/tables/impl/KeyValueTableIteratorImpl.java

```diff
@@ -0,0 +1,447 @@
+                            // Iterators are removed as they are done; if we have none left, we are all done.
+                            return CompletableFuture.completedFuture(null);
+                        }
+
+                        return populateBatch(segments);
+                    });
+        }
+
+        private CompletableFuture<IteratorItem<T>> populateBatch(PriorityQueue<PeekingIterator<T>> segments) {
+            val result = new ArrayList<T>();
+            return Futures.loop(
+                    () -> !segments.isEmpty() && result.size() < this.maxIterationSize,
+                    () -> {
+                        val first = segments.poll();
+                        result.add(first.current);
+                        return first.advance()
+                                .thenRun(() -> {
+                                    if (first.hasNext()) {
+                                        // There are still items in the segment's iterator; re-add it to the heap to add to proper position.
+                                        segments.add(first);
+                                    }
+                                });
+                    },
+                    this.executor)
+                    .thenApply(v -> new IteratorItem<>(result));
+        }
+
+        private CompletableFuture<PriorityQueue<PeekingIterator<T>>> initialize(Iterator<AsyncIterator<IteratorItem<T>>> iterators) {
+            // Fetch the first item(s) in each iterator and sort them by their first Table Key.
+            val moveFirst = new HashMap<PeekingIterator<T>, CompletableFuture<Void>>();
+            while (iterators.hasNext()) {
+                val ss = new PeekingIterator<>(iterators.next(), this.getKey);
+                moveFirst.put(ss, ss.advance());
+            }
+
+            val result = new PriorityQueue<PeekingIterator<T>>((s1, s2) -> COMPARATOR.compare(s1.currentKey, s2.currentKey));
+            return Futures.allOf(moveFirst.values())
+                    .thenApply(v -> {
+                        // Clear out those iterators with no values, and add the rest to the heap.
+                        moveFirst.keySet().stream().filter(PeekingIterator::hasNext).forEach(result::add);
+                        return result;
+                    });
+        }
+    }
+
+    //endregion
+
+    //region FlattenedAsyncIterator
+
+    /**
+     * Flattens {@link AsyncIterator} instances with a List of items into an iterator-like that returns one item at a time.
+     *
+     * @param <T> Item type.
+     */
+    @RequiredArgsConstructor
+    @VisibleForTesting
+    static class PeekingIterator<T> {
+        private final AsyncIterator<IteratorItem<T>> innerIterator;
+        private final Function<T, TableKey> getKey;
+        private volatile Iterator<T> currentBatchIterator = null;
+        @Getter
+        private volatile T current = null;
+        @Getter
+        private volatile TableKey currentKey = null;
+
```

Comment:
These three are always being set together and are all volatile this at the very least severely complicates the thread safety. Perhaps a new static inner class can be made which holds these three items immutably and then this class has a single `current` variable which is volatile and set when the iterator advances.

Reply:
I resolved this by 1) removing the iterator altogether (and having a dedicated collection with a moving index) and 2) creating a `PeekingIteratorItem` class that contains the `contents` and `contentsKey`.